### ISSUE TITLE
S3 ObjectVersion different behavior in moto

### ISSUE THREAD

**Issue by bmaisonn** [getmoto/moto#6837]
I'm using this code to recover deleted objects from AWS :

```
s3 = boto3.resource('s3')
bucket = s3.Bucket(bucket)
versions = bucket.object_versions

...


def is_delete_marker(version):
    try:
        # note head() is faster than get()
        version.head()
        return False
    except botocore.exceptions.ClientError as e:
        if 'x-amz-delete-marker' in e.response['ResponseMetadata']['HTTPHeaders']:
            return True
        # an older version of the key but not a DeleteMarker
        elif '404' == e.response['Error']['Code']:
            return False


def recover(versions, key, logger):
    for version in versions.filter(Prefix=key):
        if is_delete_marker(version):
            logger.info(f'version {version.key} {version.last_modified} recovered')
            version.delete()
```

I noticed that this code behaves differently when using moto, especially the `x-amz-delete-marker` isn't preset in `HTTPHeaders`. I didn't find a specific documentation for that but i think moto should behave the same.

**Comment by bblommers** [comment]
Hi @bmaisonn, can you share a minimal reproducible example?

What version of Moto are you using?

**Comment by bblommers** [comment]
I couldn't reproduce this based on the provided code snippet - or more specifically, I could not figure out why/when AWS returns this `x-amz-delete-marker`-header. Closing this - if you do have something reproducible I'd be happy to have another look.

**Comment by bmaisonn** [comment]
Sorry for the late answer on this @bblommers if you run this code on a bucket with versioning enabled you'll get the following output
![image](https://github.com/getmoto/moto/assets/7989822/8d78ef18-7bbe-4806-9206-2109d9d1e0a0)

In this i just create an object on s3, delete an try to remove the delete marker from the versions when there's one.

```python
import moto
import io
import boto3
import botocore


BUCKET_NAME = 'XXX'
KEY = 'test_delete_marker.txt'


s3 = boto3.resource("s3", region_name="us-east-1")
s3_client = boto3.client('s3')

with io.BytesIO() as f:
    f.write(b'test')
    s3_client.put_object(Bucket=BUCKET_NAME, Key=KEY,Body=f)

s3_client.delete_object(Bucket=BUCKET_NAME, Key=KEY)


s3 = boto3.resource('s3')
bucket = s3.Bucket(BUCKET_NAME)
versions = bucket.object_versions

def is_delete_marker(version):
    try:
        # note head() is faster than get()
        version.head()
        return False
    except botocore.exceptions.ClientError as e:
        if 'x-amz-delete-marker' in e.response['ResponseMetadata']['HTTPHeaders']:
            return True
        # an older version of the key but not a DeleteMarker
        elif '404' == e.response['Error']['Code']:
            return False


def recover(versions, key):
    for version in versions.filter(Prefix=key):
        if is_delete_marker(version):
            print(f'version {version.key} {version.last_modified} recovered')
            version.delete()

recover(versions, KEY)
```

If i run the same with moto i don't see any output here even if i see some version listed for that key

```python
import moto
import io
import boto3
import botocore


#BUCKET_NAME = 'test'
BUCKET_NAME = 'XXX'
KEY = 'test_delete_marker.txt'


with moto.mock_s3():
    s3 = boto3.resource("s3", region_name="us-east-1")
    s3_client = boto3.client('s3')

    s3.create_bucket(Bucket=BUCKET_NAME)
    s3_client.put_bucket_versioning(Bucket=BUCKET_NAME, VersioningConfiguration={'Status': 'Enabled'})


    with io.BytesIO() as f:
        f.write(b'test')
        s3_client.put_object(Bucket=BUCKET_NAME, Key=KEY,Body=f)

    s3_client.delete_object(Bucket=BUCKET_NAME, Key=KEY)


    s3 = boto3.resource('s3')
    bucket = s3.Bucket(BUCKET_NAME)
    versions = bucket.object_versions

    def is_delete_marker(version):
        try:
            # note head() is faster than get()
            version.head()
            return False
        except botocore.exceptions.ClientError as e:
            if 'x-amz-delete-marker' in e.response['ResponseMetadata']['HTTPHeaders']:
                return True
            # an older version of the key but not a DeleteMarker
            elif '404' == e.response['Error']['Code']:
                return False


    def recover(versions, key):
        for version in versions.filter(Prefix=key):
            if is_delete_marker(version):
                print(f'version {version.key} {version.last_modified} recovered')
                version.delete()

    recover(versions, KEY)
```

**Comment by bblommers** [comment]
Ah, gotcha. Not sure what I did wrong initially - but I can reproduce now. Looks like Moto returns a generic 404 when trying to call `head()` on a DeleteMarker, instead of a 405.

**Comment by jagoodhand** [comment]
@bblommers - i'm not sure the fix for this is correct.

**TLDR: In the case of an object having a delete marker - If you don't provide an object version, then a 404 error is the correct response. 405 should only be returned if you specify the version ID.**

In my case this means that moto S3 throws a different error to that produced by the real S3.

I think this fix needs to be reverted / further work required.

## Reference
The S3 documentation specifies:

> When you get an object without specifying a versionId in your request, if its current version is a delete marker, Amazon S3 responds with the following:
> 
> A 404 (Not Found) error
> 
> A response header, x-amz-delete-marker: true
> 
> When you get an object by specifying a versionId in your request, if the specified version is a delete marker, Amazon S3 responds with the following:
> 
> A 405 (Method Not Allowed) error
> 
> A response header, x-amz-delete-marker: true
> 
> A response header, Last-Modified: timestamp (only when using the [HeadObject](https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadObject.html) or [GetObject](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html) API operations)

Link: [Delete Marker User Guide](https://docs.aws.amazon.com/AmazonS3/latest/userguide/DeleteMarker.html)

**Comment by bblommers** [comment]
Appreciate the correction @jagoodhand! I've raised a PR with a fix.
